### ISSUE TITLE
Website redirects: Add Sphinx extension Rediraffe to enable doc writers to add redirects, so old doc will redirect to new doc

### ISSUE THREAD

**Issue by Coderambling** [holoviz/panel#6418]
Problem: Currently no redirects can be added when documentation files /url's change. Created this PR at the request of @maximlt

#### Describe the solution you'd like

See Discord discussion [here](https://discord.com/channels/1075331058024861767/1211959479609073714)

@maximlt  suggested using the Sphinx extension [Rediraffe](https://github.com/wpilibsuite/sphinxext-rediraffe?tab=readme-ov-file)  for this.

This will enable doc writers to use the extension to add redirect information to a new doc, so old doc will redirect to new doc.

Rediraffe can also generate auto-redirects for renamed files in the repo, although NOT for deleted files.

The autoredirect `rediraffewritediff` will have to be run regularly to look for renamed files and add them to the redirect file. So should it be run during build of the site?

This will help reduce broken pages.

This does not address showing a Not found page if a user lands on a non-existing page, for example because of a typo

#### Describe alternatives you've considered

Have not considered alternatives.

#### Additional context

See Discord discussion [here](https://discord.com/channels/1075331058024861767/1211959479609073714)

Raised this issue in Discord, created PR at request of @maximlt 

**To be addressed**

Showing a not found page with search bar.
Creation of redirects when files are deleted.

**Comment by maximlt** [comment]
Thanks for opening this issue (a Pull Request, aka PR, is a suggestion to change the content of the repository, look [there](https://github.com/holoviz/panel/pulls) to see the recent PRs) @Coderambling.

> The autoredirect rediraffewritediff will have to be run regularly to look for renamed files and add them to the redirect file. So should it be run during build of the site?

I've never used that feature of `sphinx-rediraffe`, I'm not sure how well it works and what's the best way to integrate it inn our workflow and CI.

I'd suggest starting simple:

1. Adding `sphinx-rediraffe` as a dependency of the doc build in `setup.py` and as an extension in `conf.py`
2. Using it to fix one of the identified broken links, to redirect towards a relevant page.

Once this is in place, it'll be much easier for doc writers to copy that pattern and add their own redirects when need be.

**Comment by Coderambling** [comment]
Hi @MarcSkovMadsen . Would it be better to submit things like this as a Feature Request next time, instead of as a PR? 

@maximlt suggested using sphinx-rediraffe as a solution. I have never used it myself.

Agree with your approach. Makes sense. Obviously also test this to make sure it doesn't screw up all the links or similar issues, but I guess that is part of reviewing a PR that contains code?

I do not have sufficient knowledge of Sphinx, the document system or the build system to propose actual code changes that will achieve the above.

So to implement this it needs to be assigned to someone that with the appropriate coding skills etc.

If there is a standard doc template with guidelines for writers I would suggesting adding to that with something like:

-If this doc replaces an existing doc, please add a redirect by placing the current url here, and the new url here.

**Comment by Coderambling** [comment]
Noticed a couple of links (on external sites) that do not work anymore like: https://panel.holoviz.org/user_guide/APIs.html .

Looks https://panel.holoviz.org/user_guide/ does not exist anymore (does not return a page), and therefore there could be a bunch of broken links out there that start with /user_guide ?

Proposed solution: Add a 301 that redirects /user_guide and /user_guide/* to the existing user guide main link, or at least to the Panel homepage. Can Rediraffe accomodate such a redirect, or is a different solution needed for these types of redirect?

**Comment by Coderambling** [comment]
The website has about 50-60 broken links right now, not too bad.

Should I re-check after the next update of the website when all the 1.4 docs have been added?

**Comment by maximlt** [comment]
Could you share a list of them if you have it handy? I'm also interested in how you found them.

**Comment by Coderambling** [comment]
Yes of course! Will get to it in the weekend ok?

**Comment by maximlt** [comment]
Yep no rush!

**Comment by Coderambling** [comment]
Hey @maximlt . I have attached a spreadsheet list of the 66 broken links, with an explanation. About 16 of them will be fixed with a PR @philippjfr has made. I can re-run the analysis at any time, for example after the site is renewed when 1.4 is published.

Would be happy to elaborate, as there is a broader context to this than just broken links. For example, the current site also does not have a sitemap.xml document. There really should be one.
[Panel link issues. Results of crawling panel.holoviz.org..xlsx](https://github.com/holoviz/panel/files/14715429/Panel.link.issues.Results.of.crawling.panel.holoviz.org.xlsx)

**Comment by philippjfr** [comment]
Very helpful, thanks!

**Comment by maximlt** [comment]
Ahh I thought you meant broken links of the website, i.e. pages that were moved/removed without setting up a redirect. That's not what your table contains, still, it's useful, thanks.

**Comment by Coderambling** [comment]
Sure! I could set up a weekly / monthly automated run, that logs the results in a Google Sheets, and make it shareable if that helps. Let me know.

**Comment by Coderambling** [comment]
Hey @maximlt . Ah I see what you mean.

But the table does actually contain a list of internal panel.holoviz removed pages as well.


The broken link list is used as a way to discover internal removed pages without a redirect.

So if you filter the broken links to those link that have "panel.holoviz.org/" as the domain in the URL, you have discovered a list of internal pages that have gone missing and don't have a redirect.

This does not include an "isolated" pages, that do not cause a broken link, because they are not reachable via any hyperlink on the website , but only via an internal or external search engine or something,  but that is less common.

Also, a link A on a page, that returns a 301, will redirect the user to link B.

That is fine, but ideally that link A should be replaced in the page source by link B, otherwise more and more redirects are built up on the site. It's a secundary consideration, but still. A script / filter can catch and make a table of those cases as well.

**Comment by Coderambling** [comment]
I think I remember seeing a Github PR / Issue sometime ago, that mentioned migrating away from Google Analytics to another tool for the Panel (or Param?) website. Can you point me to that Github issue and / or tell me what the new tooling is? That tooling might possibly help with the above.

**Comment by Coderambling** [comment]
Link to fixed issue regarding some external broken links https://github.com/holoviz/panel/issues/6463

**Comment by Coderambling** [comment]
Had quick look at missing links for param.holoviz.org:

696 urls in total, only 2 broken links in total (internal + external).

404 Not Found | https://param.holoviz.org/assets/param_help.png | mentioned in page: https://param.holoviz.org/user_guide/Parameters.html

404 Not Found | https://param.holoviz.org/reference.html | mentioned in page: https://param.holoviz.org/user_guide/Parameter_Types.html

**Comment by Coderambling** [comment]
Missing links for holoviews..org:

4 in total, all pointing to external missing urls:

404 Not Found | https://github.com/pyviz/aholoviews/pull/3435 | mentioned in https://holoviews.org/releases.html
404 Not Found | http://scitools.org.uk/iris/docs/v1.9.2/index.html |mentioned in https://holoviews.org/releases.html
404 Not Found | https://holoviews.org/(https://github.com/pyviz/holoviews/pull/3364 | mentioned in https://holoviews.org/releases.html
404 Not Found | https://holoviews.org/(https://github.com/pyviz/holoviews/pull/3367 | mentioned in https://holoviews.org/releases.html

So basically the page https://holoviews.org/releases.html contains 4 broken links.

**Comment by Coderambling** [comment]
Holoviz.org has 24 broken links in total, 23 external, 1 internal: https://holoviz.org/tutorial/13_Deploying_Bokeh_Apps.html

Example of an external broken link:

404 https://raw.githubusercontent.com/holoviz/holoviz/main/tutorial/05_Interactive_Pipelines.ipynbmentioned in: 

https://holoviz.org/tutorial/Interactive_Pipelines.html

If you go to the [Pipelines Tutorial page](https://holoviz.org/tutorial/Interactive_Pipelines.html) there is a section on the right-hand side that says: "Right click to download and run locally. Clicking this link will fail, because it leads to https://raw.githubusercontent.com/holoviz/holoviz/main/tutorial/05_Interactive_Pipelines.ipynb . This url doesn't exist.

So that means users can't download the .ipynb from that page. Same  problem for about10 of the other tutorials /excercise pages on holoviz.org .

**Comment by Coderambling** [comment]
Had a closer look: the /examples part is missing from the URL:


https://raw.githubusercontent.com/holoviz/holoviz/main/tutorial/05_Interactive_Pipelines.ipynb

should be:

https://raw.githubusercontent.com/holoviz/holoviz/main/examples/tutorial/05_Interactive_Pipelines.ipynb

**Comment by Coderambling** [comment]
Had a closer look: the /examples part is missing from the URL:


https://raw.githubusercontent.com/holoviz/holoviz/main/tutorial/05_Interactive_Pipelines.ipynb

should be:

https://raw.githubusercontent.com/holoviz/holoviz/main/examples/tutorial/05_Interactive_Pipelines.ipynb


Conclusion

...

What if you want to collect these pieces and put them together into a standalone app or dashboard? If so, then the next tutorial will show you how to do so!

This web page was generated from a Jupyter notebook and not all interactivity will work on this website. Right click to download and run locally for full Python-backed interactivity.

Right click to download this notebook from GitHub.

**Comment by Coderambling** [comment]
Made PR to add redirects in conf.py : https://github.com/holoviz/panel/pull/6844

**Comment by Coderambling** [comment]
See https://discord.com/channels/1075331058024861767/1361866816531468429 for further information on current broken links.
